services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -0,0 +1,204 @@
+      IDNumber: '123789',
+    };
+
+    return payload;
+  }
+
+  public async sendPaymentPerPa(
+    payload: any,
+    referenceId: string,
+  ): Promise<PaTransactionResultDto> {
+    // A timeout of 100ms to not overload safaricom server
+    await new Promise((r) => setTimeout(r, 2000));
+
+    const paTransactionResult = new PaTransactionResultDto();
+    paTransactionResult.fspName = FspName.safaricom;
+    paTransactionResult.referenceId = referenceId;
+    paTransactionResult.date = new Date();
+    paTransactionResult.calculatedAmount = payload.Amount;
+
+    const result = await this.safaricomApiService.transfer(payload);
+
+    if (result && result.ResponseCode === '0') {
+      paTransactionResult.status = StatusEnum.waiting;
+      payload.status = StatusEnum.waiting;
+    } else {
+      paTransactionResult.status = StatusEnum.error;
+      payload.status = StatusEnum.error;
+      paTransactionResult.message = result.errorMessage;
+    }
+
+    paTransactionResult.customData = {
+      requestResult: result,
+    };
+    return paTransactionResult;
+  }
+
+  public async processSafaricomRequest(
+    payload,
+    paymentRequestResultPerPa,
+  ): Promise<any> {
+    const payloadResult = Object.fromEntries(
+      Object.entries(payload).map(([k, v]) => [
+        k.charAt(0).toLowerCase() + k.slice(1),
+        v,
+      ]),
+    );
+
+    const safaricomCustomData = { ...paymentRequestResultPerPa.customData };
+    const transaction = await this.transactionRepository
+      .createQueryBuilder('transaction')
+      .where('transaction.customData ::jsonb @> :customData', {
+        customData: {
+          requestResult: {
+            ConversationID: safaricomCustomData.requestResult.ConversationID,
+          },
+        },
+      })
+      .getMany();
+
+    payloadResult.requestResult = safaricomCustomData.requestResult;
+    payloadResult.conversationID =
+      safaricomCustomData && safaricomCustomData.requestResult
+        ? safaricomCustomData.requestResult.ConversationID
+        : 'Invalid request';
+    payloadResult.transaction = transaction[0];
+    await this.safaricomRequestRepository.save(payloadResult);
+  }
+
+  public async processSafaricomResult(
+    safaricomPaymentResultData: any,
+  ): Promise<void> {
+    const safaricomDbRequest = await this.safaricomRequestRepository
+      .createQueryBuilder('safaricom_request')
+      .where('safaricom_request.requestResult ::jsonb @> :requestResult', {
+        requestResult: {
+          ConversationID: safaricomPaymentResultData.Result.ConversationID,
+        },
+      })
+      .getMany();
+
+    const transaction = await this.transactionRepository
+      .createQueryBuilder('transaction')
+      .where('transaction.customData ::jsonb @> :customData', {
+        customData: {
+          requestResult: {
+            ConversationID: safaricomPaymentResultData.Result.ConversationID,
+          },
+        },
+      })
+      .getMany();
```

Comment:
This can probably be made more efficient by using the conversationId column instead of looking in the JSON of the safaricom_request. Then you can use the found entity to get the transaction by id.

Reply:
This is fixed, now getting transaction from safaricom_request table with one query.